added autouse function scoped fixture that cleares django.contrib.sites.models.SITE_CACHE

peterlauri opened the pull request:
Closes #322

pelme commented:
Thanks for the PR!

There are some failing tests, and we also need an internal test to make sure that this fix works.

Apart from that, this PR looks good to me!

peterlauri commented:
@pelme I fixed up tests, added one test for site cache clear, should be ok to go. Please note the comment in the test, that it needs refactoring if we choose to use xdist -n option...

peterlauri commented:
Anything missing in this PR to get merged?

blueyed commented on pytest_django/plugin.py:
```diff
@@ -514,6 +514,20 @@ def _template_string_if_invalid_marker(request):
+@pytest.fixture(autouse=True, scope='function')
+def django_clear_site_cache():
+    """Clears ``django.contrib.sites.models.SITE_CACHE`` to avoid
+    unexpected behavior when cached site objects.
```
s/when/with/

blueyed commented on pytest_django/plugin.py:
```diff
@@ -514,6 +514,20 @@ def _template_string_if_invalid_marker(request):
+
+        if 'django.contrib.sites' in dj_settings.INSTALLED_APPS:
+            from django.contrib.sites import models
+            models.SITE_CACHE.clear()
```
I think that is quite some overhead for every test, especially when not using sites at all etc.

Can we create / request this fixture dynamically only if the condition is met once?

What about providing the fixture, but not automatically use it?
Even when you require it, it's not necessary for all your tests.

peterlauri replied:
I think the benefit of this autouse fixture overcomes the overhead of a list lookup, in comparison with all setup/teardown that is done with databases etc. 

We could cache the list lookup, but is that really worth it? Can we guarantee that INSTALLED_APPS are not modified by tests?

The core reason for doing this, is that the SITE_CACHE is evil :), and people that write tests for their projects where sites are used, they will tear their hair off when site objects suddenly starts to mismatch with referenced objects. The SITE_CACHE is nice for production, but not for testing...

I was against having the mailbox cleaned with an autouse fixture, as we talked about in #410. But this is a different thing, this is a cache that sits there in its evil darkness and just makes people that loves testing wanna kill them self when debugging strange errors that the SITE_CACHE causes...

pelme replied:
I think this PR is reasonable. People that are not using the site framework (not having it in INSTALLED_APPS) won't notice any difference. 

There is a helper to clear the cache `Site.objects.clear_cache()` that I think should be used instead of directly clearing the SITE_CACHE which from what I understand is not a public API:
https://docs.djangoproject.com/en/1.10/ref/contrib/sites/#caching-the-current-site-object

pelme commented on pytest_django/plugin.py:
```diff
@@ -514,6 +514,20 @@ def _template_string_if_invalid_marker(request):
 
+
+@pytest.fixture(autouse=True, scope='function')
+def django_clear_site_cache():
```
We should prefix this fixture with a `_` to make it clear that it is an internal fixture that is not meant to be requested directly by tests.

peterlauri replied:
done

pelme commented on tests/test_fixtures.py:
```diff
@@ -332,3 +332,15 @@ def test_unblock_manually(self, django_db_blocker):
+    """
+    from django.contrib.sites.models import SITE_CACHE
+    assert len(SITE_CACHE) == 0
+    SITE_CACHE[run_no] = 'run_no {}'.format(run_no)
```
From what I understand (see the comment about the `SITE_CACHE.clear()` too): SITE_CACHE is not part of a public Django API. This test should be rewritten to make sure the cache is cleared by checking `Site.objects.get_current()` which seems to be the public API to interact with the site framework.

Something like this will probably do (I have not actually run this code and probably missed some detail):
```py
@pytest.mark.django_db
@pytest.mark.parametrize('name', ['site1', 'site2'])
def test_clear_site_cache(settings, site_name):
    settings.SITE_ID = 1
    Site.objects.create(id=1, domain='foo.com', name=site_name)
    assert Site.objects.get_current().name == site_name
```

peterlauri replied:
The test is not really correct, but I get your point :) SITE_CACHE is never used if settings.SITE_ID is set, it is looked up based on hostname... but I use similar approach, unless my "unit test" that checks SITE_CACHE is enough. But it is not really enough, as in future django versions SITE_CACHE might change name or use some other way of caching...

peterlauri commented:
Btw, where do we document internal fixtures? I mean, we should describe what they do, right?

blueyed commented:
> Btw, where do we document internal fixtures? I mean, we should describe what they do, right?

Use a docstring?!

peterlauri commented:
Should we really assume that the user reads the code? My philosophy is that we should assume that all functionality is described in the public docs (readthedocs). Even that the code is documented, the normal user (read newbie), it is not sufficient. Or are we generating docs from docstrings to the public docs?

pelme commented:
Good point, we should document that we clean the sites cache but I don't think we should document "_django_clear_site_cache" explicitly in the user documentation. For that purpose a docstring is good.

We could add a section in helpers that describe additional things we do for each Django test?

peterlauri commented:
Perfect, I add a section there if magic stuff that happens in
pytest-django. Review the test, fails without the fixture, passes with it.
On Sun, 13 Nov 2016 at 14:08, Andreas Pelme <notifications@github.com>
wrote:

> Good point, we should document that we clean the sites cache but I don't
> think we should document "_django_clear_site_cache" explicitly in the user
> documentation. For that purpose a docstring is good.
>
> We could add a section in helpers that describe additional things we do
> for each Django test?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pytest-dev/pytest-django/pull/323#issuecomment-260185205>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADWiQUYZSpQH9Z9F6kzJYSQbWY2Eq3LEks5q9wu0gaJpZM4H9r-L>
> .
>

pelme commented:
Excellent! 👍 

Just prefix the fixture with `_` and add docs. :)

peterlauri commented:
@pelme added docs, changed _ and also put some notes in changelog. Done for final review.

blueyed commented:
Fails for TOXENV=py27-pytest30-django1.7-postgres (https://travis-ci.org/pytest-dev/pytest-django/jobs/175482486).

blueyed commented on docs/changelog.rst:
```diff
@@ -9,6 +9,9 @@ Features
   djangos ``mail.outbox``. The will clean/empty the ``mail.outbox`` to
   assure that no old mails are still in the outbox.
+* If ``django.contrib.sites`` is in your INSTALLED_APPS, Site cache will
+  be cleared for each tests to avoid hitting cache and cause wrong site
```
s/tests/test/

blueyed replied:
s/hitting/hitting the/

blueyed replied:
s/site/Site/

peterlauri replied:
done

blueyed commented on docs/helpers.rst:
```diff
@@ -239,3 +239,18 @@ Example
+
+pytest-django provides some pytest fixtures that are of autouse
+nature. They provide functionality to assure a clean environment
+for pytest tests.
```
s/for pytest tests/during tests/ ?!

peterlauri replied:
done

blueyed commented on docs/helpers.rst:
```diff
@@ -239,3 +239,18 @@ Example
+
+If ``django.contrib.sites`` is in your INSTALLED_APPS, Site cache will
+be cleared for each tests to avoid hitting cache and cause wrong site
+objects to be returned by ``Site.objects.get_current()``.
```
Same typos as above.

peterlauri replied:
done

peterlauri commented:
The problem is that Django==1.7 does require SITE_ID to be set if using the sites framework, so we cannot test it for the case where the request object is used to map to correct Site object.

I have added a unit test that covers Django==1.7 as well, that checks that SITE_CACHE is emptied...

blueyed commented on tests/test_environment.py:
```diff
@@ -4,9 +4,11 @@
 
 import pytest
 from django.contrib.sites.models import Site
+from django.contrib.sites.shortcuts import get_current_site
```
Unused import?

peterlauri replied:
that was fixed in e085c1b

peterlauri commented:
squashed...

pelme commented:
Thanks a lot! :)

peterlauri commented:
Nice to contribute :)
